runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
On Wed, Aug 26, 2015 at 04:05:46PM -0700, Saied Kazemi wrote:
> Re /proc/<pid>/fd/<fd>, this information is available only for
> opened files and the paths will point inside the container
> namespace.

I can confirm this when accessed from the host:

  # lsof -p 20050 +E
  COMMAND   PID  USER   FD   TYPE DEVICE SIZE/OFF     NODE NAME
  …
  exe     20050 wking    0r   REG   0,19   140520 10149337 /lib64/ld-2.20.so
  …

where I'm looking at that file in my container's mount namespace, not
the host's.  So if we go with the /proc approach, users that want
host-side paths will need to unpack the mount namespacing to find the
host-side path.

> Also, not all of the external paths remain open during the entire
> lifetime of the container.

Why would this matter?  You're using CRIU to checkpoint/restore the
container as it stands, not for some pre-checkpoint state.

> If this information is not provided in the runtime state
> configuration file, it will have to be looked up via
> /proc/<pid>/mountinfo.

That seems like the only reasonable approach to me, but I can see sane
people differing on who does that direct-kernel lookup (i.e. the
runtime or the state consumer).  If you're just caching
container-creation-time config settings, you'll miss any subsequent
mounts that may have been made since creation time (e.g. via hooks
[1]).

[1]: https://github.com/opencontainers/specs/pull/87#discussion_r38043014